btcutil/block.go

```diff
@@ -154,12 +154,26 @@ func (b *Block) Transactions() []*Tx {
 		b.transactions = make([]*Tx, len(b.msgBlock.Transactions))
 	}
 
+	// Offset of each tx.  80 accounts for the block header size.
+	offset := 80 + wire.VarIntSerializeSize(uint64(len(b.msgBlock.Transactions)))
+
 	// Generate and cache the wrapped transactions for all that haven't
 	// already been done.
 	for i, tx := range b.transactions {
 		if tx == nil {
 			newTx := NewTx(b.msgBlock.Transactions[i])
 			newTx.SetIndex(i)
+
+			size := b.msgBlock.Transactions[i].SerializeSize()
```

Comment:
IIUC, with this change, we only get the benefit when `block.Transactions()` is called, while https://github.com/btcsuite/btcd/pull/1376 will ensure the serialization is cached as soon as things are read off the wire. The one thing this has that the other PR doesn't is storing the serialization for both segwit and non-segwit transactions. Doing this actually simplifies the logic a bit, as then you can just tee the entire thing and skip the segwit specific parts for the non-segwit encoding.

Reply:
Also this isn't addressing the same issue as #1376 is, see this profile: https://github.com/btcsuite/btcd/pull/1376#issuecomment-1467126518 

When we go to calculate the sighash we want to hash the entire serialization, which isn't the same thing as a normal dsha256 to compute the `txid`: https://github.com/btcsuite/btcd/blob/7520523a9a54aff213f5455b048a55f818118a84/txscript/sighash.go#L173